content/en/altinity-kb-integrations/ClickHouse_python_drivers.md

```diff
@@ -0,0 +1,152 @@
+
+[Best practices with flask · Issue #73 · ClickHouse/clickhouse-connect](https://github.com/ClickHouse/clickhouse-connect/issues/73#issuecomment-1325280242)
+
+#### clickhouse-connect & clickhouse-driver with Asyncio
+
+`clickhouse-connect` code is synchronous and running synchronous functions in an async application is a workaround and might not be as efficient as using a library designed for asynchronous operations from the ground up. Problem is there are few libs/modules in Python. So you can use `concurrent.futures` and `ThreadpoolExecutor` or `ProcessPoolExecutor`. Python GIL has a mutex over Threads but not to Processes so if you need performance at the cost of using processes instead of threads (not much different for medium workloads) you can use `ProcesspoolExecutor` instead.
+
+Some info about this from the tinybird guys [https://www.tinybird.co/blog-posts/killing-the-processpoolexecutor](https://www.tinybird.co/blog-posts/killing-the-processpoolexecutor)
+
+For clickhouse-connect
+
+```python
+import asyncio
+from concurrent.futures import ThreadPoolExecutor
+import clickhouse_connect
+
+# Function to execute a query using clickhouse-connect synchronously
+def execute_query_sync(query):
+    client = clickhouse_connect.get_client()  # Adjust connection params as needed
+    result = client.query(query)
+    return result
+
+# Asynchronous wrapper function to run the synchronous function in a thread pool
+async def execute_query_async(query):
+    loop = asyncio.get_running_loop()
+    # Use ThreadPoolExecutor to execute the synchronous function
+    with ThreadPoolExecutor() as pool:
+        result = await loop.run_in_executor(pool, execute_query_sync, query)
+        return result
+
+async def main():
+    query = "SELECT * FROM your_table LIMIT 10"  # Example query
+    result = await execute_query_async(query)
+    print(result)
+
+# Run the async main function
+if __name__ == '__main__':
+    asyncio.run(main())
+```
+
+Clickhouse-driver code is also synchronous and suffers the same problem as clickhouse-connect
+
+[https://clickhouse-driver.readthedocs.io/en/latest/quickstart.html#async-and-multithreading](https://clickhouse-driver.readthedocs.io/en/latest/quickstart.html#async-and-multithreading)
+
+So to use asynchronous approach it is recommended to use a connection pool and some asyncio wrapper that can hide the complexity of using the `ThreadPoolExecutor/ProcessPoolExecutor`
+
+To begin testing such environment [aiohttp](https://docs.aiohttp.org/) is a good approach. Here an example:
+
+[https://github.com/lesandie/clickhouse-tests/blob/main/scripts/test_aiohttp_inserts.py](https://github.com/lesandie/clickhouse-tests/blob/main/scripts/test_aiohttp_inserts.py)
+
+This will use simply requests module and aiohttp (you can tune the connection pooler [https://docs.aiohttp.org/en/stable/client_advanced.html#limiting-connection-pool-size](https://docs.aiohttp.org/en/stable/client_advanced.html#limiting-connection-pool-size))
```

Comment:
not a python guy, but does this make sense? "This simply uses the `requests` module and `aiohttp` ...

Reply:
Yes not well expressed I'll rewrite